src/classlibnative/bcltype/number.cpp

```diff
@@ -354,21 +354,39 @@ static unsigned DigitsToInt(__in_ecount(count) wchar* p, int count)
 //
 // multiply two numbers in the internal integer representation
 //
-static UINT64 Mul64Lossy(UINT64 a, UINT64 b, INT* pexp)
+static UINT64 Mul64x64To128Upper(UINT64 a, UINT64 b, INT* pexp)
 {
     LIMITED_METHOD_CONTRACT
 
-    // it's ok to losse some precision here - Mul64 will be called
-    // at most twice during the conversion, so the error won't propagate
-    // to any of the 53 significant bits of the result
-    UINT64 val = Mul32x32To64(a >> 32, b >> 32) +
-        (Mul32x32To64(a >> 32, b) >> 32) +
-        (Mul32x32To64(a, b >> 32) >> 32);
+#if defined(_MSC_VER) && defined(_TARGET_AMD64_)
+    uint64_t upper = __umulh(a, b);
```

Comment:
Can we use this on Unix also?
https://github.com/llvm-mirror/clang/blob/9e45fbd56cdcc6c62a276b5838878cc65dee4831/lib/Headers/intrin.h#L292

Reply:
I dont think that section is included for unix. It looks like it's under the MSC_VER section

Reply:
Does it simply compile down to MUL anyway? How is it better than `a * b` ?

Reply:
Because `a * b` returns a 64-bit result. We need the upper 64-bits of the 128-bit result, which is what `__umulh` gives. 

Under the hood, the intrinsic just calls mul (which already computes a 128-bit result for x64) and returns the upper bits (which are in the `rdx` register).

Reply:
Ah right. Would surprise me if there isn't a Linux equivalent intrinsic for such a useful operation.

Reply:
`(int64_t)(((__int128_t)x * y) >> 64)` will do the right thing here on any recent clang or gcc.